Add introduction to CDEvents (#34)
cdevents/spec: README.md

It's an incubated project at the
[Continuous Delivery Foundation](https://cd.foundation) (CDF).

## Background
<!--
Resources used for the background text
https://cd.foundation/blog/2021/03/16/cd-foundation-announces-industry-initiative-to-standardize-events-from-ci-cd-systems/
https://github.com/cdfoundation/toc/blob/master/sigs/sig-events.md
https://github.com/cdfoundation/sig-events
https://github.com/cdfoundation/toc/blob/master/proposals/cdevents/cdevents.md
-->
In a complex and fast moving CI/CD world with a lot of different tools and
platforms that need to communicate with each other interoperability stands as a
crucial thing. The maintainer of a CI/CD system needs to swap out tools in short
time with little to no stops.

The larger and more complex a CI/CD system becomes, challenges increase in
knowing how the tools communicate and what they do.

### What we provide

> The CDEvents protocol defines a vocabulary of events enabling tools to

Reviewer: As this is the first time "events" is mentioned in this document (which is probably the first document people read on CDEvents), I believe we should either link to a description of what an event is here, or add a sentence/paragraph about it inline

It's an incubated project at the
[Continuous Delivery Foundation](https://cd.foundation) (CDF).

## Background
<!--
Resources used for the background text
https://cd.foundation/blog/2021/03/16/cd-foundation-announces-industry-initiative-to-standardize-events-from-ci-cd-systems/
https://github.com/cdfoundation/toc/blob/master/sigs/sig-events.md
https://github.com/cdfoundation/sig-events
https://github.com/cdfoundation/toc/blob/master/proposals/cdevents/cdevents.md
-->
In a complex and fast moving CI/CD world with a lot of different tools and
platform that need to communicate with each other interoperability stands as a
crucial thing. The maintainer of a CI/CD system needs to swap out tools in short
time with little to no stops.

The larger and more complex a CI/CD system becomes, challenges increase in
knowing how the tools communicate and what they do.

### What we provide

The CDEvents protocol defines a vocabulary of events enabling tools to
communicate in a interoperable way.

We extend other efforts such as CloudEvents by introducing purpose and semantics

> meaning of an event.

Reviewer: ```suggestion
to the event.
```

We thus build on top of the work from other players as the following stack

> pictures exemplifies.

Author: Updated in 1ad4def

> ![stack](./images/stack.png)

Reviewer 2: It looks like something went wrong with the rendering of this diagram. The font looks rather jagged...

Author: Hmm, that's not good. I however don't see the problem do you have some more info?

Reviewer 2: ok no worries, I don't want to block the PR on this - I see you included the source as well so I can also try to re-render it on my side

> By providing a interoperable way of tool to communicate we also provide means to

Reviewer: ```suggestion
By providing an interoperable way of tools to communicate we also provide means to
```

> give an overview picture increasing observability, but also provides measuring

Reviewer: ```suggestion
give an overview picture increasing observability, but also to give measuring
```

points for metrics.

> #### How does it enable tools to communicate in a interoperable way?

Reviewer 2: Perhaps from this point on we could add this material to the `primer.md` document?

By creating a language, we define how tools and services communicate with each
other about occurrences in a CI/CD system. As this language does not tie to a
specific tool it serves a neutral ground for communication.

> Using this language we defining a set of events that has purpose and semantic

Reviewer: ```suggestion
Using this language we define a set of events with purpose and semantic
```

interpret the information received. This enables tools to have a common
understanding of the information sent in the events.

The language enables creating an ecosystem of tools for monitoring, tracing,
measuring, and orchestrating using our events without having to write a
"plugin" for every tool.

#### Why use events?

Reading from the
[CloudEvents primer - design goals](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/primer.md#design-goals)

> The goal of the CloudEvents specification is to define interoperability of
> event systems that allow services to produce or consume events, where the
> producer and consumer can be developed and deployed independently. A producer
> can generate events before a consumer is listening, and a consumer can express
> an interest in an event or class of events that is not yet being produced.

We believe that using events will lead to a more decoupled systems with services
and tools developed and deployed independently. This makes us agnostic of the
underlying infrastructure

#### Why not point-to-point communication?

We believe that using integrations based on point-to-point communication will
create a system that will:

> * Not scale - when trying to add new consumers or senders each tools have to make

Reviewer: ```suggestion
* Not scale - when trying to add new consumers or producers each tool have to make
```

  an update
* Create a coupled architecture - using point-to-point communication creates a
  tightly intertwined architecture difficult to expand and monitor.

> #### Why not webhooks?

Reviewer 2: I'm not sure about this. I see webhooks as a way to deliver events, and I didn't know we recommended explicitly against them. CloudEvents include a spec for webhooks https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/http-webhook.md and they could be an interesting option in cases where authorization is required.

Author: Felt this was a bit on the edge so I removed it in 1ad4def
